lib/phoenix/channel.ex

```diff
@@ -137,14 +137,18 @@ defmodule Phoenix.Channel do
                                                                   :ignore |
                                                                   {:error, reason :: term, Socket.t}
 
-  defcallback leave(msg :: map, Socket.t) :: {:ok, Socket.t}
+  defcallback leave(msg :: map, Socket.t) :: :ok | {:error, reason :: term}
 
   defcallback handle_in(event :: String.t, msg :: map, Socket.t) :: {:ok, Socket.t} |
```

Comment:
Are we keeping `:ok | :leave | :error` for backwards compatibility here for now? Also, we may want to consider supporting `{:leave, reason, reply, state}` and `{:leave, reason, state}` as in GenServer (maybe we should just rename it to :stop and go with terminate as mentioned above).